libraries/AP_HAL_ChibiOS/Util.cpp

```diff
@@ -86,16 +86,11 @@ void Util::free_type(void *ptr, size_t size, AP_HAL::Util::Memory_Type mem_type)
 
 void *Util::allocate_heap_memory(size_t size)
 {
-    void *buf = malloc(size);
-    if (buf == nullptr) {
+    memory_heap_t *heap = (memory_heap_t *)malloc(size + sizeof(memory_heap_t));
+    if (heap == nullptr) {
         return nullptr;
     }
-
-    memory_heap_t *heap = (memory_heap_t *)malloc(sizeof(memory_heap_t));
-    if (heap != nullptr) {
-        chHeapObjectInit(heap, buf, size);
-    }
-
+    chHeapObjectInit(heap, heap + 1U, size);
```

Comment:
```suggestion
    chHeapObjectInit(heap, heap + sizeof(memory_heap_t), size);
```

Don't just add 1, that may or may not be the size of a `memory_heap_t` a quick inspection seems to show that a `memory_heap_t `is much larger than that, and you are probably corrupting the heap.

Reply:
Ah, yes, I'm sorry. I managed to confuse myself from the diff, where it had been `void *` and the fact that the signature of `chHeapObjectInit` is actually looking for `void *`. Given the implicit cast to `void *` I think it's still more clear as adding the size, but I'm pretty sure you are actually correct. Sorry about that!